### PR TITLE
Refactor utils for getting 'params'

### PR DESCRIPTION
**Pull request by sifnoc**
In this PR, Two issues covered https://github.com/summa-dev/summa-solvency/issues/100 and https://github.com/summa-dev/summa-solvency/issues/89

For trust setup, now we only use `generate_setup_artifacts`.

```Rust
pub fn generate_setup_artifacts<C: Circuit<Fp> + CircuitExt<Fp>>(
    k: u32,
    params_path: Option<&str>,
    circuit: C,
) -> Result<
    (
        ParamsKZG<Bn256>,
        ProvingKey<G1Affine>,
        VerifyingKey<G1Affine>,
    )
```

This PR will be merged later than #116

### CODE REVIEW

File: zk_prover/src/circuits/utils.rs [hunk 1]
```diff
@@ -33,11 +33,32 @@ use std::{
...
+    circuit: C,
+) -> Result<
+    (
+        ParamsKZG<Bn256>,
+        ProvingKey<G1Affine>,
+        VerifyingKey<G1Affine>,
+    ),
+    &'static str,
+> {
+    let mut ptau_path = format!("ptau/hermez-raw-{}", k);
```
Comment (enricobottazzi) [inline zk_prover/src/circuits/utils.rs:54]: Can we remove this line and any reference to `ptau_path`? I think that the `generate_setup_artifacts` function should only consider two cases: 

- Case in which no `params_path` is provided as input. In that case the trusted setup is generated randomly.
- Case in which `params_path` is provided as input. In that case the trusted setup is generated starting from the ptau file from `params_path`

File: zk_prover/src/circuits/utils.rs [hunk 2]
```diff
@@ -47,14 +68,38 @@ pub fn generate_setup_params(k: u32) -> ParamsKZG<Bn256> {
         let timer = start_timer!(|| "Creating params");
         let params = ParamsKZG::<Bn256>::setup(k, OsRng);
         end_timer!(timer);
-        params
+        if params.k() < k {
```
Comment (enricobottazzi) [inline zk_prover/src/circuits/utils.rs:71]: In which case this error would get triggered here? 

While I see this error being possible in the `else` case. Here it seems to me that the `.k()` of params will always be equal to the `k` that we passed as input as `params` are generated using `k`.

File: zk_prover/src/circuits/utils.rs (hunk 1 above)
Comment (enricobottazzi) [reply zk_prover/src/circuits/utils.rs:54]: If you think that this modification makes sense, please also update the description on how to generate the trusted setup in the README

File: zk_prover/benches/full_solvency_flow.rs [hunk 3]
```diff
@@ -62,11 +62,10 @@ fn verification_key_gen_mst_inclusion_circuit(_c: &mut Criterion) {
 fn proving_key_gen_mst_inclusion_circuit(_c: &mut Criterion) {
     let mut criterion = Criterion::default().sample_size(SAMPLE_SIZE);
 
-    let params: ParamsKZG<Bn256> = generate_setup_params(13);
-
     let empty_circuit = MstInclusionCircuit::<LEVELS, L, N_ASSETS>::init_empty();
 
-    let vk = keygen_vk(&params, &empty_circuit).expect("vk generation should not fail");
+empty_circuit.clone()let (params, _, vk) = generate_setup_artifacts(13, None, empty_circuit.clone()).unwrap();
```
Comment (enricobottazzi) [inline zk_prover/benches/full_solvency_flow.rs:67]: I think there's a typo here

File: zk_prover/benches/full_solvency_flow.rs [hunk 4]
```diff
@@ -1,15 +1,15 @@
 use criterion::{criterion_group, criterion_main, Criterion};
 use halo2_proofs::{
-    halo2curves::bn256::{Bn256, Fr as Fp},
-    plonk::{keygen_pk, keygen_vk},
+    halo2curves::bn256::{Bn256, Fr as Fp, G1Affine},
```
Comment (enricobottazzi) [inline zk_prover/benches/full_solvency_flow.rs:3]: There's some unused import here

**Review by enricobottazzi — CHANGES_REQUESTED** [review]

File: zk_prover/src/circuits/utils.rs (hunk 2 above)
Comment (sifnoc) [reply zk_prover/src/circuits/utils.rs:71]: True, no need this check on that. I will remove it.

File: zk_prover/src/circuits/utils.rs (hunk 1 above)
Comment (sifnoc) [reply zk_prover/src/circuits/utils.rs:54]: Totally makes sense. I was wrong in checking the `ptau` file in that function.

File: zk_prover/benches/full_solvency_flow.rs (hunk 3 above)
Comment (sifnoc) [reply zk_prover/benches/full_solvency_flow.rs]: Fixed at [42b1dba](https://github.com/summa-dev/summa-solvency/pull/117/commits/42b1dbad462239eb315390299bf33e84148aa294)

File: zk_prover/benches/full_solvency_flow.rs (hunk 4 above)
Comment (sifnoc) [reply zk_prover/benches/full_solvency_flow.rs]: Fixed at [42b1dba](https://github.com/summa-dev/summa-solvency/pull/117/commits/42b1dbad462239eb315390299bf33e84148aa294)

File: zk_prover/src/circuits/utils.rs (hunk 2 above)
Comment (sifnoc) [reply zk_prover/src/circuits/utils.rs]: Fixed at [42b1dba](https://github.com/summa-dev/summa-solvency/pull/117/commits/42b1dbad462239eb315390299bf33e84148aa294)

File: zk_prover/src/circuits/utils.rs (hunk 1 above)
Comment (sifnoc) [reply zk_prover/src/circuits/utils.rs]: Fixed at [42b1dba](https://github.com/summa-dev/summa-solvency/pull/117/commits/42b1dbad462239eb315390299bf33e84148aa294)

File: zk_prover/src/circuits/utils.rs [hunk 5]
```diff
@@ -27,35 +33,48 @@ use snark_verifier::{
...
-        let timer = start_timer!(|| "Creating params");
-        let mut params_fs = File::open(ptau_path).expect("couldn't load params");
-        let params = ParamsKZG::<Bn256>::read(&mut params_fs).expect("Failed to read params");
-        end_timer!(timer);
-        params
+/// Generate setup artifacts for a circuit of size `k`, where 2^k represents the number of rows in the circuit.
+///
+/// If the trusted setup parameters are not found, the function performs an unsafe trusted setup to generate the necessary parameters
+/// If the provided `k` value is larger than the `k` value of the loaded parameters, an error is returned, as the provided `k` is too large.
+/// Otherwise, if the `k` value is smaller than the `k` value of the loaded parameters, the parameters are downsized to fit the requested `k`.
```
Comment (enricobottazzi) [inline zk_prover/src/circuits/utils.rs:41]: I see that here you claim that you are downsizing the params but I don't see it being implmented in the function

Comment (sifnoc) [reply zk_prover/src/circuits/utils.rs:41]: Ah It missed out, Fixed at [9ab6c78](https://github.com/summa-dev/summa-solvency/pull/117/commits/9ab6c786f1e4961dbfc7922b366257653bf236bd)

**Review by enricobottazzi — APPROVED** [review]

**Pull request merged**